### PR TITLE
Add instrumentation of driver_requests_* metrics

### LINKED ISSUE #152: Implement newly added "driver_requests" metrics
**What would you like to be added**:
In version [v0.50.0](https://github.com/gardener/machine-controller-manager/releases/tag/v0.50.0) of MCM new metrics were added to record driver request durations and failures. It is for the provider's to implement the logic of updating these metrics at the appropriate points in the code base based on their driver activity with the provider.  

**Why is this needed**:

In previous versions of MCM with the in-tree provider, we relied on metrics such as `mcm_cloud_api_requests_failed_total` to visualise any failures from the underlying AWS requests and diagnose problems.  In the latest version of MCM and the aws provider we do not have this metric anymore. 

Implementing `driver_request_duration_seconds` and `driver_requests_failed_total` will provide visibility again on the underlying calls being made to AWS.

### PR DESCRIPTION
**Pull request by saley89**
**What this PR does / why we need it**:
In version [v0.50.0](https://github.com/gardener/machine-controller-manager/releases/tag/v0.50.0) of MCM new metrics were added to record driver request durations and failures. It is for the provider's to implement the logic of updating these metrics at the appropriate points in the code base based on their driver activity with the provider.  

This PR is based on the [machine-controller-manager-provider-azure](https://github.com/gardener/machine-controller-manager-provider-azure/blob/d4b3b5d4d48cf3ea95bd7e39b334aeb3016e95d2/pkg/azure/instrument/instrument.go#L62-L82) implementation of these metrics and adds the duration of requests and records any failures during these operations:

* create machine
* delete machine
* list machines
* get machine status 
* get volume ids

**Which issue(s) this PR fixes**:
Fixes #152

**Special notes for your reviewer**:

See the linked `machine-controller-manager-provider-azure` project and whether we are able to take their `instrument.go` function as is.

**Release note**:
```feature user
Implements the driver metrics added to MCM in version `0.50.0` such that duration of calls to AWS and any failed requests are recorded:
* driver_request_duration_seconds
* driver_requests_failed_total
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @saley89 for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by rishabh-11** [comment]
Hey @saley89, with mcm v0.50 we also have introduced metrics for cloud provider API calls. I have raised an issue (https://github.com/gardener/machine-controller-manager-provider-aws/issues/156) for implementing these in provider-aws.  Would you like to implement these metrics as a part of this PR or wait for the MCM team to implement it when we pick it up?

**Comment by saley89** [comment]
Hi @rishabh-11 my team has a need for the metrics that are implemented by this change and I noticed it was relatively straight forward to add these in based on the implementation in the other provider.  

If you are comfortable for this PR to go in for these metrics without the others that would be beneficial for us.  The other metrics could go in as and when you have time/are able to do so.

**Review by rishabh-11 — CHANGES_REQUESTED** [review]
Can you move the `instrument.go` file in a separate directory called `instrument` inside `pkg`?

**Comment by saley89** [comment]
@rishabh-11 sure, that is done now.

File: pkg/instrument/instrument.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		)
+		if errors.As(err, &statusErr) {
+			labels = append(labels, strconv.Itoa(int(statusErr.Code())))
+		}
+		metrics.DriverFailedAPIRequests.
+			WithLabelValues(labels...).
+			Inc()
+		return
+	}
+	// compute the time taken to complete the AZ service call and record it as a metric
```
Comment (rishabh-11) [inline pkg/instrument/instrument.go:30]:
```suggestion
	// compute the time taken to complete the AWS service call and record it as a metric
```

**Review by rishabh-11 — CHANGES_REQUESTED** [review]
just a minor change. rest is good. Also did you test this? Are the metrics working as expected?

**Comment by saley89** [comment]
Regarding tests I think it can be done.  Looks like the way to go about it would be to leverage the testutils package of the client_go/prometheus package we already have in this repo.  Some of example of using it are in the [prometheus repo](https://github.com/prometheus/client_golang/blob/main/prometheus/testutil/testutil_test.go).

I had a brief look this afternoon to see if we can introduce it to the existing `core_test.go` tests but would need more time to get something ready for the PR.  In the meantime anyone can contribute to it if they have a solution. 

My thinking would be ideally to get the metric, call methods like `CreateMachine`, `DeleteMachine` and see that the metric with the relevant labels has indeed been incremented. Unfortunately the Azure example hasn't tested it either for reference.

**Comment by rishabh-11** [comment]
Hey @saley89, While testing your PR I found a bug in the code related to metrics. Can you refer to https://github.com/gardener/machine-controller-manager-provider-azure/pull/130 and fix this PR?

**Comment by saley89** [comment]
@rishabh-11 I was able to pull in the changes needed for the updated driver API function. I needed to revendor due to go mod changes to bring in the latest prometheus golang client.  

I will rebase with master to fix these conflicts.

**Comment by saley89** [comment]
The rebase has got things back in order and this is now ready for your review again. Thanks

**Review by unmarshall — CHANGES_REQUESTED** [review]

File: pkg/aws/core.go [hunk 2]
```diff
@@ -63,7 +72,9 @@ func NewAWSDriver(spi spi.SessionProviderInterface) driver.Driver {
 }
 
 // CreateMachine handles a machine creation request
-func (d *Driver) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (*driver.CreateMachineResponse, error) {
+func (d *Driver) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (resp *driver.CreateMachineResponse, err error) {
+	defer instrument.DriverAPIMetricRecorderFn(createMachineOperationLabel, &err)
```
[the remark this answers is not in the document]
Comment (saley89) [reply pkg/aws/core.go]: Fixed

**Review by rishabh-11 — APPROVED** [review]
/lgtm

**Review by unmarshall — APPROVED** [review]
/lgtm

**Pull request merged**